actionview/lib/action_view/helpers/text_helper.rb

```diff
@@ -139,15 +139,15 @@ def highlight(text, phrases, options = {}, &block)
         if text.blank? || phrases.blank?
           text || ""
         else
-          match = Array(phrases).map do |p|
+          pattern = Array(phrases).map do |p|
             Regexp === p ? p.to_s : Regexp.escape(p)
           end.join("|")
 
           if block_given?
-            text.gsub(/(#{match})(?![^<]*?>)/i, &block)
+            do_highlight(text, pattern, &block)
           else
             highlighter = options.fetch(:highlighter, '<mark>\1</mark>')
-            text.gsub(/(#{match})(?![^<]*?>)/i, highlighter)
+            do_highlight(text, pattern, highlighter: highlighter)
           end
```

Comment:
Another alternative:

```suggestion
          patterns = Array(phrases).map { |phrase| Regexp === phrase ? phrase : Regexp.escape(phrase) }
          pattern = /(#{patterns.join("|")})/i
          highlighter = options.fetch(:highlighter, '<mark>\1</mark>') unless block

          text.scan(/<[^>]+>|[^<]+/).each do |segment|
            if !segment.start_with?("<")
              if block
                segment.gsub!(pattern, &block)
              else
                segment.gsub!(pattern, highlighter)
              end
            end
          end.join
```



Benchmark:

```ruby
# frozen_string_literal: true
require "benchmark/ips"

TestController = Class.new(ActionController::Base)

def benchmark(input, pattern)
  puts "=== input length: #{input.length} ".ljust(72, "=")

  Benchmark.ips do |x|
    x.report("original") do
      TestController.helpers.highlight(input, pattern)
    end

    x.report("commit 204fac2") do
      TestController.helpers.highlight_204fac2(input, pattern)
    end

    x.report("alternative") do
      TestController.helpers.highlight_alt(input, pattern)
    end

    x.compare!
  end
end

benchmark("foo bar 1234\n" * 10000, /foo|bar|1234/)
benchmark("You searched for: rails", /for|rails/)
```

```console
$ bin/rails r pr47339.rb
=== input length: 130000 ===============================================
Warming up --------------------------------------
            original     1.000  i/100ms
      commit 204fac2     1.000  i/100ms
         alternative     1.000  i/100ms
Calculating -------------------------------------
            original      0.017  (± 0.0%) i/s -      1.000  in  58.022911s
      commit 204fac2      0.413  (± 0.0%) i/s -      3.000  in   7.277325s
         alternative     20.534  (± 4.9%) i/s -    103.000  in   5.028064s

Comparison:
         alternative:       20.5 i/s
      commit 204fac2:        0.4 i/s - 49.77x  (± 0.00) slower
            original:        0.0 i/s - 1191.45x  (± 0.00) slower

=== input length: 23 ===================================================
Warming up --------------------------------------
            original   367.000  i/100ms
      commit 204fac2   200.000  i/100ms
         alternative   329.000  i/100ms
Calculating -------------------------------------
            original      3.776k (±14.2%) i/s -     18.717k in   5.068264s
      commit 204fac2      1.857k (±21.7%) i/s -      9.000k in   5.059772s
         alternative      2.676k (±13.3%) i/s -     13.160k in   5.012414s

Comparison:
            original:     3776.2 i/s
         alternative:     2675.8 i/s - 1.41x  (± 0.00) slower
      commit 204fac2:     1856.8 i/s - 2.03x  (± 0.00) slower
```

Reply:
The full code without the wonky diff highlighting:

```ruby
def highlight(text, phrases, options = {}, &block)
  text = sanitize(text) if options.fetch(:sanitize, true)

  if text.blank? || phrases.blank?
    text || ""
  else
    patterns = Array(phrases).map { |phrase| Regexp === phrase ? phrase : Regexp.escape(phrase) }
    pattern = /(#{patterns.join("|")})/i
    highlighter = options.fetch(:highlighter, '<mark>\1</mark>') unless block

    text.scan(/<[^>]+>|[^<]+/).each do |segment|
      if !segment.start_with?("<")
        if block
          segment.gsub!(pattern, &block)
        else
          segment.gsub!(pattern, highlighter)
        end
      end
    end.join
  end.html_safe
end
```

Reply:
👏 👏 👏 

I was thinking initially about trying `scan`, but haven't tried it for some reason.

Added you as a co-author (should be author actually 😄)